File: OOP/OOP_variable/set_get_1.py
class Cal(object):
    def __init__(self,v1,v2):
        self.v1 = v1 #인스턴스 변수 v1,v2 선언
        self.v2 = v2

    def add(self):
        return self.v1+self.v2 #인스턴스 변수 v1+v2

    def sub(self):
        return self.v1-self.v2 #인스턴스 변수 v1-v2
    
    def setV1(self, v): #setV1은 self.v1을 int 값으로 설정
        if isinstance(v, int):
            self.v1 = v
        
    def getV1(self): #getV1은 self.v1의 값을 가져온다.
        return self.v1

c1 = Cal(10,10)
print(c1.add()) # 20
print(c1.sub()) # 0

c1.setV1('one') # 문자열 'one'
c1.v2 = 30 # 숫자 30 
print(c1.add())
print(c1.sub())
# 문자열과 숫자를 계산하려 해서 error

File: OOP/OOP_variable/set_get_1.rb
class Cal
    attr_reader:v1, :v2 #get
    attr_writer:v1 #set
    #attr_accessor 호출

    def initialize(v1,v2)
        @v1 = v1 # 인스턴스 변수
        @v2 = v2
    end
    
    def add()
        return @v1+@v2
    end
    
    def sub()
        return @v1-@v2
    end

    def setV1(v)
        if v.is_a?(Integer) # .is_a?() 인스턴스가 어떤 타입인지 알아볼때 쓰는 메서드
            @v1 = v
        end
    end

    def getV1()
        return @v1
    end
end

c1 = Cal.new(10,10)
p c1.add()
p c1.sub()

c1.setV1('one') # setV1에서 Integer가 아니기 때문에 문자열인 V1을 무시한다.
c1.v1 = 20
p c1.add()
p c1.sub()
